File: src/test/java/com/robohorse/robopojogenerator/generator/processing/ClassProcessorTest.java
package com.robohorse.robopojogenerator.generator.processing;

import com.robohorse.robopojogenerator.generator.common.ClassField;
import com.robohorse.robopojogenerator.generator.common.ClassItem;
import com.robohorse.robopojogenerator.generator.common.JsonItem;
import com.robohorse.robopojogenerator.generator.utils.ClassGenerateHelper;
import org.json.JSONArray;
import org.json.JSONObject;
import org.junit.Before;
import org.junit.Test;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.MockitoAnnotations;
import testutils.JsonReader;

import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

import static org.junit.Assert.*;
import static org.mockito.Mockito.when;

/**
 * Created by vadim on 02.10.16.
 */

public class ClassProcessorTest {
    private JsonReader jsonReader = new JsonReader();

    @InjectMocks
    ClassProcessor classProcessor;
    @Mock
    ClassGenerateHelper classGenerateHelper;

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);
    }

    @Test
    public void testSingleObjectGeneration_isCorrect() throws Exception {
        final JSONObject jsonObject = jsonReader.read("single_object.json");
        final String name = "Response";

        when(classGenerateHelper.formatClassName(name))
                .thenReturn(name);

        final Map<String, ClassItem> classItemMap = new HashMap<>();
        final JsonItem jsonItem = new JsonItem(jsonObject, name);
        classProcessor.proceed(jsonItem, classItemMap);
        assertTrue(classItemMap.size() == 1);

        Iterator iterator = classItemMap.values().iterator();
        ClassItem classItem = (ClassItem) iterator.next();
        assertEquals(classItem.getClassName(), name);

        final Map<String, ClassField> fields = classItem.getClassFields();
        assertNotNull(fields);

        for (String key : jsonObject.keySet()) {
            assertTrue(fields.containsKey(key));
        }
    }

    @Test
    public void testInnerObjectGeneration_isCorrect() throws Exception {
        final JSONObject jsonObject = jsonReader.read("inner_json_object.json");
        final JSONObject innerJsonObject = jsonObject.optJSONObject("data");
        final String name = "Response";

        when(classGenerateHelper.formatClassName(name))
                .thenReturn(name);

        final Map<String, ClassItem> classItemMap = new HashMap<>();
        final JsonItem jsonItem = new JsonItem(jsonObject, name);
        classProcessor.proceed(jsonItem, classItemMap);
        assertTrue(classItemMap.size() == 2);

        for (ClassItem classItem : classItemMap.values()) {
            final Map<String, ClassField> fields = classItem.getClassFields();
            assertNotNull(fields);

            if (name.equalsIgnoreCase(classItem.getClassName())) {
                for (String key : jsonObject.keySet()) {
                    assertTrue(fields.containsKey(key));
                }
            } else {
                for (String key : innerJsonObject.keySet()) {
                    assertTrue(fields.containsKey(key));
                }
            }
        }
    }

    @Test
    public void testEmptyArrayGeneration_isCorrect() throws Exception {
        final JSONObject jsonObject = jsonReader.read("empty_array.json");
        final String name = "Response";

        when(classGenerateHelper.formatClassName(name))
                .thenReturn(name);
        final Map<String, ClassItem> classItemMap = new HashMap<>();
        final JsonItem jsonItem = new JsonItem(jsonObject, name);

        classProcessor.proceed(jsonItem, classItemMap);
        assertTrue(classItemMap.size() == 1);

        Iterator iterator = classItemMap.values().iterator();
        ClassItem classItem = (ClassItem) iterator.next();
        assertEquals(classItem.getClassName(), name);

        final Map<String, ClassField> fields = classItem.getClassFields();
        assertNotNull(fields);

        for (String key : jsonObject.keySet()) {
            assertTrue(fields.containsKey(key));
        }

        final ClassField targetObjectType = classItem.getClassFields().get("data");

        assertEquals("List<Object>", targetObjectType.getJavaItem());
    }

    @Test
    public void testIntegerArrayGeneration_isCorrect() throws Exception {
        final JSONObject jsonObject = jsonReader.read("array_with_primitive.json");
        final String name = "Response";
        final String targetType = "List<Integer>";

        when(classGenerateHelper.formatClassName(name))
                .thenReturn(name);

        final Map<String, ClassItem> classItemMap = new HashMap<>();
        final JsonItem jsonItem = new JsonItem(jsonObject, name);
        classProcessor.proceed(jsonItem, classItemMap);

        assertTrue(classItemMap.size() == 1);

        Iterator iterator = classItemMap.values().iterator();
        ClassItem classItem = (ClassItem) iterator.next();
        assertEquals(classItem.getClassName(), name);

        final Map<String, ClassField> fields = classItem.getClassFields();
        assertNotNull(fields);

        for (String key : jsonObject.keySet()) {
            assertTrue(fields.containsKey(key));
        }

        final ClassField actualType = classItem.getClassFields().get("data");

        assertEquals(targetType, actualType.getJavaItem());
    }

    @Test
    public void testInnerArrayObjectGeneration_isCorrect() throws Exception {
        final JSONObject jsonObject = jsonReader.read("array_with_jsonobject.json");
        final JSONArray innerJsonArray = jsonObject.optJSONArray("data");
        final JSONObject innerJsonObject = innerJsonArray.getJSONObject(0);
        final String name = "Response";
        final String targetType = "List<DataItem>";

        when(classGenerateHelper.formatClassName(name))
                .thenReturn(name);
        when(classGenerateHelper.getClassNameWithItemPostfix(Mockito.anyString()))
                .thenReturn("DataItem");

        final Map<String, ClassItem> classItemMap = new HashMap<>();
        final JsonItem jsonItem = new JsonItem(jsonObject, name);
        classProcessor.proceed(jsonItem, classItemMap);

        assertTrue(classItemMap.size() == 2);

        for (ClassItem classItem : classItemMap.values()) {
            final Map<String, ClassField> fields = classItem.getClassFields();
            assertNotNull(fields);

            if (name.equalsIgnoreCase(classItem.getClassName())) {
                for (String key : jsonObject.keySet()) {
                    assertTrue(fields.containsKey(key));
                    final ClassField actualType = classItem.getClassFields().get("data");
                    final String javaItem = actualType.getJavaItem();
                    assertEquals(targetType, javaItem);
                }
            } else {
                for (String key : innerJsonObject.keySet()) {
                    assertTrue(fields.containsKey(key));
                }
            }
        }
    }
}

File: src/test/kotlin/com/robohorse/robopojogenerator/generator/processing/ClassProcessorTest.kt
package com.robohorse.robopojogenerator.generator.processing

import com.robohorse.robopojogenerator.generator.consts.common.ClassItem
import com.robohorse.robopojogenerator.generator.consts.common.JsonModel.JsonItem
import com.robohorse.robopojogenerator.generator.utils.ClassGenerateHelper
import com.robohorse.robopojogenerator.utils.JsonReader
import io.mockk.every
import io.mockk.impl.annotations.InjectMockKs
import io.mockk.impl.annotations.RelaxedMockK
import io.mockk.junit5.MockKExtension
import org.json.JSONObject
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import java.util.*
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

@ExtendWith(MockKExtension::class)
class ClassProcessorTest {
    private val jsonReader: JsonReader = JsonReader()

    @InjectMockKs
    lateinit var classProcessor: ClassProcessor

    @RelaxedMockK
    lateinit var classGenerateHelper: ClassGenerateHelper

    @Test
    fun testSingleObjectGeneration_isCorrect() {
        val jsonObject = jsonReader.read("single_object.json") as JSONObject
        val name = "Response"
        every { classGenerateHelper.formatClassName(name) }.returns(name)
        val classItemMap: Map<String, ClassItem> = HashMap()
        val jsonItem = JsonItem(name, jsonObject)
        classProcessor.proceed(jsonItem, classItemMap)
        assertTrue(classItemMap.size == 1)
        val iterator = classItemMap.values.iterator()
        val (className, _, _, _, fields) = iterator.next()
        assertEquals(className, name)
        assertNotNull(fields)
        for (key in jsonObject.keySet()) {
            assertTrue(fields.containsKey(key))
        }
    }

    @Test
    fun testInnerObjectGeneration_isCorrect() {
        val jsonObject = jsonReader.read("inner_json_object.json") as JSONObject
        val innerJsonObject = jsonObject.optJSONObject("data")
        val name = "Response"
        every { classGenerateHelper.formatClassName(name) }.returns(name)

        val classItemMap: Map<String, ClassItem> = HashMap()
        val jsonItem = JsonItem(name, jsonObject)
        classProcessor.proceed(jsonItem, classItemMap)
        assertTrue(classItemMap.size == 2)
        for ((className, _, _, _, fields) in classItemMap.values) {
            assertNotNull(fields)
            if (name.equals(className, ignoreCase = true)) {
                for (key in jsonObject.keySet()) {
                    assertTrue(fields.containsKey(key))
                }
            } else {
                for (key in innerJsonObject.keySet()) {
                    assertTrue(fields.containsKey(key))
                }
            }
        }
    }

    @Test
    fun testEmptyArrayGeneration_isCorrect() {
        val jsonObject = jsonReader.read("empty_array.json") as JSONObject
        val name = "Response"
        every { classGenerateHelper.formatClassName(name) }.returns(name)
        val classItemMap: Map<String, ClassItem> = HashMap()
        val jsonItem = JsonItem(name, jsonObject)
        classProcessor.proceed(jsonItem, classItemMap)
        assertTrue(classItemMap.size == 1)
        val iterator: Iterator<*> = classItemMap.values.iterator()
        val (className, _, _, _, fields) = iterator.next() as ClassItem
        assertEquals(className, name)
        assertNotNull(fields)
        for (key in jsonObject.keySet()) {
            assertTrue(fields.containsKey(key))
        }
        val targetObjectType = fields["data"]
        assertEquals("List<Object>", targetObjectType!!.getJavaItem())
    }

    @Test
    fun testIntegerArrayGeneration_isCorrect() {
        val jsonObject = jsonReader.read("array_with_primitive.json") as JSONObject
        val name = "Response"
        val targetType = "List<Integer>"
        every { classGenerateHelper.formatClassName(name) }.returns(name)
        val classItemMap: Map<String, ClassItem> = HashMap()
        val jsonItem = JsonItem(name, jsonObject)
        classProcessor.proceed(jsonItem, classItemMap)
        assertTrue(classItemMap.size == 1)
        val iterator: Iterator<*> = classItemMap.values.iterator()
        val (className, _, _, _, fields) = iterator.next() as ClassItem
        assertEquals(className, name)
        assertNotNull(fields)
        for (key in jsonObject.keySet()) {
            assertTrue(fields.containsKey(key))
        }
        val actualType = fields["data"]
        assertEquals(targetType, actualType!!.getJavaItem())
    }

    @Test
    fun testInnerArrayObjectGeneration_isCorrect() {
        val jsonObject = jsonReader.read("array_with_jsonobject.json") as JSONObject
        val innerJsonArray = jsonObject.optJSONArray("data")
        val innerJsonObject = innerJsonArray.getJSONObject(0)
        val name = "Response"
        val targetType = "List<DataItem>"
        every { classGenerateHelper.formatClassName(name) }.returns(name)
        every { classGenerateHelper.getClassNameWithItemPostfix(any()) }.returns("DataItem")
        val classItemMap: Map<String, ClassItem> = HashMap()
        val jsonItem = JsonItem(name, jsonObject)
        classProcessor.proceed(jsonItem, classItemMap)
        assertTrue(classItemMap.size == 2)
        for ((className, _, _, _, fields) in classItemMap.values) {
            assertNotNull(fields)
            if (name.equals(className, ignoreCase = true)) {
                for (key in jsonObject.keySet()) {
                    assertTrue(fields.containsKey(key))
                    val actualType = fields["data"]
                    val javaItem = actualType!!.getJavaItem()
                    assertEquals(targetType, javaItem)
                }
            } else {
                for (key in innerJsonObject.keySet()) {
                    assertTrue(fields.containsKey(key))
                }
            }
        }
    }
}
